src/docker-extract.c

```diff
@@ -0,0 +1,320 @@
+/* extract_tar
+ *  Extract a tar file to rootfs_dir using libarchive. Handles compression.
+ *  Exclude any .wh. whiteout files, and device/pipe/fifo entries
+ *
+ * See https://github.com/libarchive/libarchive/wiki/Examples#A_Complete_Extractor
+ */
+int extract_tar(const char *tarfile, const char *rootfs_dir) {
+    int retval = 0;
+    struct archive *a;
+    struct archive *ext;
+    struct archive_entry *entry;
+    int flags;
+    int r;
+    char *orig_dir;
+
+    orig_dir = get_current_dir_name();
+
+    /* Select which attributes we want to restore. */
+    flags = ARCHIVE_EXTRACT_TIME;
+    flags |= ARCHIVE_EXTRACT_PERM;
+    flags |= ARCHIVE_EXTRACT_ACL;
+    flags |= ARCHIVE_EXTRACT_FFLAGS;
+
+    a = archive_read_new();
+    archive_read_support_format_all(a);
+#if ARCHIVE_VERSION_NUMBER <= 3000000
+    archive_read_support_compression_all(a);
+#else
+    archive_read_support_filter_all(a);
+#endif
+    ext = archive_write_disk_new();
+    archive_write_disk_set_options(ext, flags);
+    archive_write_disk_set_standard_lookup(ext);
+    if ((r = archive_read_open_filename(a, tarfile, 10240))){
+        singularity_message(ERROR, "Error opening tar file %s\n", tarfile);
+        ABORT(255);
+    }
+
+    // Extract into the SINGULARITY_ROOTFS
+    r = chdir(rootfs_dir);
+    if (r < 0 ){
+        singularity_message(ERROR, "Could not chdir to SINGULARITY_ROOTFS %s\n", rootfs_dir);
+        ABORT(255);
+    }
+
+    for (;;) {
+        r = archive_read_next_header(a, &entry);
+
+        if (r == ARCHIVE_EOF)
+            break;
+
+        if (r < ARCHIVE_OK)
+            singularity_message(WARNING, "Warning reading tar header: %s\n", archive_error_string(a));
+        if (r < ARCHIVE_WARN) {
+            singularity_message(ERROR, "Error reading tar header: %s\n", archive_error_string(a));
+            ABORT(255);
+        }
+
+        const char *pathname = archive_entry_pathname(entry);
+        int pathtype = archive_entry_filetype(entry);
+
+        // Do not extract whiteout markers (handled in apply_whiteouts)
+        // Do not extract sockers, chr/blk devices, pipes
+        if (strstr(pathname, "/.wh.") || pathtype == AE_IFSOCK ||
+            pathtype == AE_IFCHR || pathtype == AE_IFBLK || pathtype == AE_IFIFO) {
+            continue;
+        }
+
+        r = archive_write_header(ext, entry);
+        if (r < ARCHIVE_OK)
+            singularity_message(WARNING, "Warning handling tar header: %s\n", archive_error_string(ext));
+        else if (archive_entry_size(entry) > 0) {
+            r = copy_data(a, ext);
+            if (r < ARCHIVE_OK)
+                singularity_message(WARNING, "Warning handling tar header: %s\n", archive_error_string(ext));
+            if (r < ARCHIVE_WARN) {
+                singularity_message(ERROR, "Error handling tar header: %s\n", archive_error_string(ext));
+                ABORT(255);
+            }
+        }
+        r = archive_write_finish_entry(ext);
+        if (r < ARCHIVE_OK)
```

Comment:
Sorry, meant to include this in the previous review.  I'm going to advocate that we always use curly braces for if statements even if they are just simple one liners like this.  Otherwise my brain skips a beat when I'm reading and I have to stop and do a double take. lol